### PR TITLE
add a --dot option so we can get an actual graph

### PR DESCRIPTION
**Pull request by ylmrx**
##### SUMMARY

This implement a "dot graphviz" output for `ansible-inventory`. This is very similar to "`terraform graph`" command.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

ansible-inventory

##### ADDITIONAL INFORMATION

I wonder if the vars options is relevant here, they are hard to make it pleasing in a usable graphic as soon as you go beyond really trivial inventory (we use _huge_ ones here).
Also my color choices are probably not so good.
Please advise.

```paste below
# consider this inventory
$ cat /tmp/trivial_inventory
[group1]
host1
host2

[group2]
host3

[group3]
host3
host4

[machines:children]
group1
group2
# the basic graph option
$ ansible-inventory -i /tmp/trivial_inventory --graph                           
@all:
  |--@group3:
  |  |--host3
[...]
  |  |--@group2:
  |  |  |--host3
  |--@ungrouped:
# the new --dot option
$ ansible-inventory -i /tmp/trivial_inventory --dot | dot -T svg -o inv.svg -x

 [attachment converted to png below this code block, because github can't embed svg]

$ ansible-inventory -i /tmp/trivial_inventory --dot
digraph G {
graph [rankdir="LR" bgcolor="#F6F6F6"];
node [shape="record" color="#D3D3D3" style="filled" fillcolor="#AAAAAA" contraint="false"];
edge [color="#4284F3"];
    "d87c448044defb778f33158d8ccf94a20531d600" [label="all"];
        "e157bec2c6b2adc3ce70aca06ae4f6dfc1ee91ee" [label="group3"];
        "3b39992c0b1a539a76f9b80752690a1ca0b77c59" [label="host3"];
    "e157bec2c6b2adc3ce70aca06ae4f6dfc1ee91ee" -> "3b39992c0b1a539a76f9b80752690a1ca0b77c59";
        "abf67cf30541723b65dd6b9fc7c0b1caf0101fcf" [label="host4"];
    "e157bec2c6b2adc3ce70aca06ae4f6dfc1ee91ee" -> "abf67cf30541723b65dd6b9fc7c0b1caf0101fcf";
  "d87c448044defb778f33158d8ccf94a20531d600" -> "e157bec2c6b2adc3ce70aca06ae4f6dfc1ee91ee";
        "6516c232528481e757d3d0eec7e42b291b27aca6" [label="machines"];
            "be4f69bf25177a377d20437445e49a13cf3c65d5" [label="group1"];
            "e5e2a9518607915ae99ab77d575bfe7a7dcf2a99" [label="host1"];
      "be4f69bf25177a377d20437445e49a13cf3c65d5" -> "e5e2a9518607915ae99ab77d575bfe7a7dcf2a99";
            "e0d87a849f5a1e3d140e8b666446536edfe92089" [label="host2"];
      "be4f69bf25177a377d20437445e49a13cf3c65d5" -> "e0d87a849f5a1e3d140e8b666446536edfe92089";
    "6516c232528481e757d3d0eec7e42b291b27aca6" -> "be4f69bf25177a377d20437445e49a13cf3c65d5";
            "ca5ff4060839997ae17316da3095bc65d342f828" [label="group2"];
            "3b39992c0b1a539a76f9b80752690a1ca0b77c59" [label="host3"];
      "ca5ff4060839997ae17316da3095bc65d342f828" -> "3b39992c0b1a539a76f9b80752690a1ca0b77c59";
    "6516c232528481e757d3d0eec7e42b291b27aca6" -> "ca5ff4060839997ae17316da3095bc65d342f828";
  "d87c448044defb778f33158d8ccf94a20531d600" -> "6516c232528481e757d3d0eec7e42b291b27aca6";
        "3b36d99c7304a08a5a24f247f2e47e6b283f147e" [label="ungrouped"];
  "d87c448044defb778f33158d8ccf94a20531d600" -> "3b36d99c7304a08a5a24f247f2e47e6b283f147e";
}
```
example_inv

### CODE REVIEW

File: lib/ansible/cli/inventory.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 
 import argparse
 from operator import attrgetter
+from hashlib import sha1
```
Comment (felixfontein) [inline lib/ansible/cli/inventory.py]: I guess this can be removed. You import sha1 locally below.

**Review by felixfontein — COMMENTED** [review]
You need a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

File: lib/ansible/cli/inventory.py (hunk 1 above)
Comment (ylmrx) [reply lib/ansible/cli/inventory.py]: fixed in 7b1e546

File: lib/ansible/cli/inventory.py [hunk 2]
```diff
@@ -277,6 +281,47 @@ def inventory_graph(self):
...
+        return result
+
+    def inventory_graph_dot(self):
+
+        start_at = self._get_group(context.CLIARGS['pattern'])
+        if context.CLIARGS['show_vars']:
+            raise AnsibleOptionsError("This would be graphic")
+
+        output = [
+            'digraph G {',
```
Comment (bcoca) [inline lib/ansible/cli/inventory.py:325]: option should deal with the case the dependencies are not avaiable and give a warning.

File: lib/ansible/cli/inventory.py [hunk 3]
```diff
@@ -277,6 +281,47 @@ def inventory_graph(self):
...
+            'graph [rankdir="LR" bgcolor="#F6F6F6"];',
+            'node [shape="record" color="#D3D3D3" style="filled" fillcolor="#AAAAAA" contraint="false"];',
+            'edge [color="#4284F3"];',
```
Comment (bcoca) [inline lib/ansible/cli/inventory.py]: colors should all be configurable

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/cli/inventory.py [hunk 4]
```diff
@@ -77,16 +77,18 @@ def init_parser(self):
         action_group.add_argument("--list", action="store_true", default=False, dest='list', help='Output all hosts info, works as inventory script')
         action_group.add_argument("--host", action="store", default=None, dest='host', help='Output specific host info, works as inventory script')
         action_group.add_argument("--graph", action="store_true", default=False, dest='graph',
-                                  help='create inventory graph, if supplying pattern it must be a valid group name')
+                                  help='Create inventory graph, if supplying pattern it must be a valid group name')
+        action_group.add_argument("--dot", action="store_true", default=False, dest='graph_dot',
```
Comment (bcoca) [inline lib/ansible/cli/inventory.py:81]: should we ask for a file name? (default to stdout is fine)

File: lib/ansible/cli/inventory.py [hunk 5]
```diff
@@ -277,6 +281,47 @@ def inventory_graph(self):
...
+                new_parent_group = '%s"%s" -> "%s";' % (
+                    ' ' * depth * 2, hashed_group_name, hashed_host_name)
+                result.append(new_parent_group)
+        return result
+
+    def inventory_graph_dot(self):
+
+        start_at = self._get_group(context.CLIARGS['pattern'])
+        if context.CLIARGS['show_vars']:
+            raise AnsibleOptionsError("This would be graphic")
```
Comment (sivel) [inline lib/ansible/cli/inventory.py]: This should probably just be done with a mutually exclusive group in the CLI argparse declaration.

File: lib/ansible/cli/inventory.py (hunk 4 above)
Comment (ylmrx) [reply lib/ansible/cli/inventory.py:81]: I don't think it's relevant for this change as other "formats" options don't have such behavior.
I'd rather prepare another MR to implement a "-o" parameter and allow any of the options (graph, json, list, ...) to have an output file. wdyt ?

File: lib/ansible/cli/inventory.py (hunk 2 above)
Comment (ylmrx) [reply lib/ansible/cli/inventory.py:325]: (ditto from irc, but best to have it here)
What dependency do you have in mind ? the only import I added is hashlib, it doesn't add any deps (It's not using the "graphviz" python module at all if it's what you assumed)

File: lib/ansible/cli/inventory.py (hunk 3 above)
Comment (ylmrx) [reply lib/ansible/cli/inventory.py]: fixed in ba475655

File: lib/ansible/cli/inventory.py (hunk 5 above)
Comment (ylmrx) [reply lib/ansible/cli/inventory.py]: is this what you had in mind ? cf 629dcc6bf4

File: lib/ansible/cli/inventory.py (hunk 4 above)
Comment (ylmrx) [reply lib/ansible/cli/inventory.py:81]: disregard previous comment. --output is already implemented (silly me).
it just work (--output FILE, and if no output provided, default to stdout)

File: lib/ansible/cli/inventory.py (hunk 2 above)
Comment (ylmrx) [reply lib/ansible/cli/inventory.py:325]: graphviz dependency signaled in the template.

**Comment by bcoca** [comment]
related #72499

**Comment by s-hertel** [comment]
Hi!

Thanks very much for your submission to Ansible.  It means a lot to us that you've taken time to contribute.

Unfortunately, we're not sure if we want this feature in the program, and I don't want this to seem confrontational.
Our reasons for this are:

   * We'd like to add a support dot graphs more generically, like #72499.

However, we're absolutely always up for discussion.
Because this project is very active, we're unlikely to see comments made on closed tickets and we lock them after some time.
If you or anyone else has any further questions, please let us know by using any of the communication methods listed in the page below:

   * https://docs.ansible.com/ansible/latest/community/communication.html

In the future, sometimes starting a discussion on the development list prior to implementing a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!

**Comment by ylmrx** [comment]
The linked PR makes much more sense I'll try to come up with a nice j2 template for this :)

**Pull request closed without merging**